tokio/src/fs/write.rs

```diff
@@ -24,8 +24,59 @@ use std::{io, path::Path};
 /// # }
 /// ```
 pub async fn write(path: impl AsRef<Path>, contents: impl AsRef<[u8]>) -> io::Result<()> {
-    let path = path.as_ref().to_owned();
+    let path = path.as_ref();
     let contents = crate::util::as_ref::upgrade(contents);
 
+    #[cfg(all(tokio_uring, feature = "rt", feature = "fs", target_os = "linux"))]
+    {
+        let handle = crate::runtime::Handle::current();
+        let driver_handle = handle.inner.driver().io();
+        if driver_handle.check_and_init()? {
+            return write_uring(path, contents).await;
+        }
+    }
+
+    write_spawn_blocking(path, contents).await
+}
+
+#[cfg(all(tokio_uring, feature = "rt", feature = "fs", target_os = "linux"))]
+async fn write_uring(path: &Path, mut buf: OwnedBuf) -> io::Result<()> {
+    use crate::{fs::OpenOptions, runtime::driver::op::Op};
+    use std::os::fd::OwnedFd;
+
+    let file = OpenOptions::new()
+        .write(true)
+        .create(true)
+        .truncate(true)
+        .open(path)
+        .await?;
+
+    let mut fd: OwnedFd = file
+        .try_into_std()
+        .expect("unexpected in-flight operation detected")
+        .into();
+
+    let total: usize = buf.len();
+    let mut offset: usize = 0;
+    while offset < total {
+        // There is a cap on how many bytes we can write in a single uring write operation.
+        // ref: https://github.com/axboe/liburing/discussions/497
+        let len = std::cmp::min(total - offset, u32::MAX as usize) as u32;
```

Comment:
Could also make sense to move this logic inside `Op::write_at` and have it accept usize. After all, it's indistinguishable to the caller from any other situation that writes the bytes partially. That way it's more natural for the caller as the caller expects usize for both input and output length.
```suggestion
        let len = u32::try_from(total - offset).unwrap_or(u32::MAX);
```